tx/textual/valuerenderer/coin_test.go

```diff
@@ -0,0 +1,75 @@
+package valuerenderer_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	bankv1beta1 "cosmossdk.io/api/cosmos/bank/v1beta1"
+	basev1beta1 "cosmossdk.io/api/cosmos/base/v1beta1"
+	"cosmossdk.io/tx/textual/valuerenderer"
+)
+
+// mockCoinMetadataKey is used in the mock coin metadata querier.
+func mockCoinMetadataKey(denom string) string {
+	return fmt.Sprintf("%s-%s", "coin-metadata", denom)
+}
+
+// mockCoinMetadataQuerier is a mock querier for coin metadata used for test
+// purposes.
+func mockCoinMetadataQuerier(ctx context.Context, denom string) (*bankv1beta1.Metadata, error) {
+	v := ctx.Value(mockCoinMetadataKey(denom))
+	if v == nil {
+		return nil, nil
+	}
+
+	return v.(*bankv1beta1.Metadata), nil
+}
+
+func TestFormatCoin(t *testing.T) {
+	var testcases []coinTest
+	raw, err := ioutil.ReadFile("../internal/testdata/coin.json")
+	require.NoError(t, err)
+	err = json.Unmarshal(raw, &testcases)
+	require.NoError(t, err)
+
+	textual := valuerenderer.NewTextual(mockCoinMetadataQuerier)
+
+	for _, tc := range testcases {
+		metadata := &bankv1beta1.Metadata{
+			Display:    tc.metadata.Denom,
+			DenomUnits: []*bankv1beta1.DenomUnit{{Denom: tc.coin.Denom, Exponent: 0}, {Denom: tc.metadata.Denom, Exponent: tc.metadata.Exponent}},
+		}
+		ctx := context.WithValue(context.Background(), mockCoinMetadataKey(tc.coin.Denom), metadata)
+
+		r, err := textual.GetValueRenderer(fieldDescriptorFromName("COIN"))
+		require.NoError(t, err)
+		b := new(strings.Builder)
+		err = r.Format(ctx, protoreflect.ValueOf(tc.coin.ProtoReflect()), b)
+		require.NoError(t, err)
```

Comment:
Add test cases for each error exit from the renderer so you get 100% coverage.

Also use the parse method (which should be implemented for this PR) to prove round-tripping.

Reply:
Added test cases to bring file coverage to 89%. However 100% coverage seems like a futile exercice, lines like [this one](https://github.com/cosmos/cosmos-sdk/blob/4670093c567959ed3928b1c4776dea642c6bf26f/tx/textual/valuerenderer/coins.go#L129) are not worth the effort to cover.

Added an inline comment about parsing, see https://github.com/cosmos/cosmos-sdk/pull/12729#discussion_r962963732